src/platform/w3dengine/client/gui/guicallbacks/w3dcontrolbar.cpp

```diff
@@ -0,0 +1,565 @@
+
+                    if (start_y - i5 > 0) {
+                        g_theDisplay->Set_Clip_Region(&clip_region);
+                        g_theWindowManager->Win_Draw_Image(
+                            centerBar, screen_x, i5, end_x, i5 + centerBar->Get_Image_Height(), 0xFFFFFFFF);
+                        g_theDisplay->Enable_Clipping(false);
+                    }
+
+                    g_theWindowManager->Win_Draw_Image(
+                        beginBar, screen_x, start_y, screen_x + width, screen_y + height, 0xFFFFFFFF);
+                    g_theWindowManager->Win_Draw_Image(endBar,
+                        screen_x,
+                        screen_y + height - i4,
+                        screen_x + width,
+                        screen_y + height - end_bar_height - i4,
+                        0xFFFFFFFF);
+                } else {
+                    g_theWindowManager->Win_Draw_Image(beginBar, screen_x, start_y, end_x, screen_y + height, 0xFFFFFFFF);
+                    g_theWindowManager->Win_Draw_Image(endBar,
+                        screen_x,
+                        screen_y + height - end_bar_height - begin_bar_height,
+                        width + screen_x,
+                        screen_y + height - begin_bar_height,
+                        0xFFFFFFFF);
+                }
+            }
+        }
+    }
+}
+
+void W3D_Command_Bar_Top_Draw(GameWindow *window, WinInstanceData *instance)
+{
+    GameWindow *win = g_theWindowManager->Win_Get_Window_From_Id(
+        nullptr, g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:ButtonGeneral"));
+
+    if (win != nullptr) {
+        if (!win->Win_Is_Hidden()) {
+            g_thePlayerList->Get_Local_Player()->Is_Player_Active();
+        }
+    }
+}
+
+void W3D_Command_Bar_Background_Draw(GameWindow *window, WinInstanceData *instance)
+{
+    ControlBarSchemeManager *manager = g_theControlBar->Get_Control_Bar_Scheme_Manager();
+
+    if (manager != nullptr) {
+        static const NameKeyType winNamekey = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:BackgroundMarker");
+        GameWindow *win = g_theWindowManager->Win_Get_Window_From_Id(nullptr, winNamekey);
+
+        if (win != nullptr) {
+            static ICoord2D basePos;
+            g_theControlBar->Get_Background_Marker_Pos(&basePos.x, &basePos.y);
+            int screen_x;
+            int screen_y;
+            win->Win_Get_Screen_Position(&screen_x, &screen_y);
+            ICoord2D offset;
+            offset.y = screen_y - basePos.y;
+            offset.x = screen_x - basePos.x;
+            manager->Draw_Background(offset);
+        }
+    }
+}
+
+void W3D_Command_Bar_Foreground_Draw(GameWindow *window, WinInstanceData *instance)
+{
+    ControlBarSchemeManager *manager = g_theControlBar->Get_Control_Bar_Scheme_Manager();
+
+    if (manager != nullptr) {
+        static const NameKeyType winNamekey = g_theNameKeyGenerator->Name_To_Key("ControlBar.wnd:BackgroundMarker");
```

Comment:
Not ForegroundMarker? There is a `ControlBar.wnd:ForegroundMarker` in wnd file.

Reply:
No, this is correct.